casanovo/denovo/model.py

```diff
@@ -408,21 +875,27 @@ def on_predict_epoch_end(
             return
         for batch in results:
             for step in batch:
-                for spectrum_i, precursor, peptide, aa_scores in zip(*step):
-                    peptide = peptide[1:]
-                    peptide_tokens = re.split(r"(?<=.)(?=[A-Z])", peptide)
-                    # Take the scores of the most probable amino acids.
-                    top_aa_scores = torch.max(
-                        aa_scores[1 : len(peptide_tokens) + 1], axis=1
-                    )[0]
-                    peptide_score = torch.mean(top_aa_scores).detach().item()
+                for spectrum_i, precursor, aa_tokens, aa_scores in zip(*step):
+                    # Omit stop token.
+                    aa_tokens = aa_tokens[1:]
+                    peptide = "".join(aa_tokens)
+                    # If this is a non-finished beam (after exceeding
+                    # `max_length`), return a dummy (empty) peptide.
+                    if len(peptide) == 0:
```

Comment:
**question:** I don't understand these two lines. `peptide` is created from `aa_tokens` and only has length 0 if `aa_tokens` is empty, so this sets it from an empty list to an empty string? Both are empty iterables, so this code doesn't really do anything?

Reply:
If `aa_tokens` has a number of empty strings `len(aa_tokens)>0` which we try to iterate over below. The list of empty strings arise from dummy tokens.

Reply:
Got it, thanks for the explanation. Maybe we should add a comment with that info for future reference.

However, I think that the code will crash in some of the following lines in that case. For example, if `aa_tokens` is empty, `top_aa_scores` will also be an empty list, and taking the mean when calculating the peptide score will crash.

So I don't think we have a test for this situation yet? And maybe rather than setting `aa_scores` to empty, we should not do any of the following code and directly give a 0-score to the mzTab writer?

Reply:
Can you also verify that the AA scores and peptide sequence are correctly ordered? Right now we take the peptide from the `aa_tokens` directly in line 881, but we ultimately reverse the AA scores before writing them in line 925. It seems to me that those two operations are not in sync with each other.

Reply:
re: empty list, as `np.mean` returns NaN without crashing the method doesn't crash and outputs empty string for the peptide and NaN for score in the mzTab. I think this can stay as is but I can implement a form of zero output if you think that's better.

Reply:
re: AA scores and tokens order, I corrected the bug introduced in [053967e](https://github.com/Noble-Lab/casanovo/pull/87/commits/053967edc80b31c998dcb446c7aae861d04b3fd8), refactored the code and added tests for this functionality.

Reply:
Woopsie. Thanks for catching that!

I was also too fast in checking `np.mean`, which indeed doesn't crash on empty input but gives a `RuntimeWarning`. I saw a stack trace and didn't look closely enough. But this works then.